Improve CompareMDWorkspaces for comparing MDEvents (master)

wdzhou opened the pull request:
**Description of work.**

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

Improve the MDEvent comparison in algorithm CompareMDWorkspaces.

This is ORNL [SPEC #105](https://code.ornl.gov/sns-hfir-scse/spectroscopy/spectroscopy/-/issues/105)

**To test:**

<!-- Instructions for testing. -->

Fixes #xxxx. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

wdzhou committed:
Prototype sorting of MDEvents in a box.

wdzhou committed:
Added a copy constructor.

wdzhou committed:
Fixed issues for MacOS.

wdzhou committed:
Fixed cppcheck issues.

wdzhou committed:
Enable auto tolerane on float comparison.

ckendrick committed:
Initial equation update to IkedaCarpenterPV docs

ckendrick committed:
Move parameters to table and minor doc updates

wdzhou committed:
Merge remote-tracking branch 'origin/ornl-next' into SPEC94_imporve_compare_md

wdzhou committed:
Refs #105.  Better comparison workflow on events in a box.

wdzhou committed:
Refs #105.  Improve event comparison error message.

wdzhou committed:
Refs #105.  Fix issue with comparing centers.

wdzhou committed:
Refs #105.  Tidy up.

wdzhou committed:
Refs #105.  Rename a method.

jmborr commented:
@wdzhou Is this ready for review or are you planning on doing more commits?

wdzhou committed:
Refs #105. No evaluaton on MDGridBox.

KedoKudo committed:
remove tolerance check for optimization results

KedoKudo committed:
remove tolerance setting from API

KedoKudo committed:
update release doc

jmborr commented on Framework/MDAlgorithms/src/CompareMDWorkspaces.cpp:
```diff
@@ -31,6 +32,153 @@ class CompareFailsException : public std::runtime_error {
+      // equal is hard to tell
+      coord_t diff = mCoordinates[i] - event2.mCoordinates[i];
+
+      if (fabs(mCoordinates[i]) > 1E-7 && fabs(diff / mCoordinates[i]) < 1E-6) {
```
Class `SimpleMDEvent` should have a `double static SimpleMDEvent::tolerance` attribute which can be set by algorithm `CompareMDWorkspaces` and default to `1E-07`.

wdzhou replied:
float is good enough.  

The `static` is good idea.

jmborr commented on Framework/MDAlgorithms/src/CompareMDWorkspaces.cpp:
```diff
@@ -210,28 +363,83 @@ void CompareMDWorkspaces::compareMDWorkspaces(typename MDEventWorkspace<MDE, nd>
+
+            // convert MDEvents vectors to SimpleMDEvent vectors for comparison
             for (size_t i = 0; i < events1.size(); i++) {
+              std::vector<float> centers1;
```
I think it runs faster if youallocate `std::vector<float> centers1(nd)` and then assign `centers1[d] = events1[i].getCenter(d)`

wdzhou replied:
Good point.  I will make test to see the difference.

jmborr commented on Framework/MDAlgorithms/src/CompareMDWorkspaces.cpp:
```diff
@@ -31,6 +32,153 @@ class CompareFailsException : public std::runtime_error {
+        // larger than epsilon and relative difference less than epsilon
+        equal = true;
+        less = false;
+      } else if (fabs(mCoordinates[i]) < 1E-7 && diff < 1E-7) {
```
Should be `fabs(diff) < 1E-7`

wdzhou replied:
I forgot the `fabs`

jmborr commented on Framework/MDAlgorithms/src/CompareMDWorkspaces.cpp:
```diff
@@ -31,6 +32,153 @@ class CompareFailsException : public std::runtime_error {
+
+  float getError() const { return mError; }
+
+  bool operator()(SimpleMDEvent &event1, const SimpleMDEvent &event2) { return event1 < event2; }
```
Why are you allowing `SimpleMDEvent` instances to be called?

wdzhou replied:
This is a requirement by std::sort.

jmborr replied:
Can you point to me where's that requirement? I can sort a vector of `int` objects with `std::sort` and the `int` object doesn't have the `()` operator.

wdzhou replied:
it is because integer has its own `()` operator.  You don't need it for a structure either.  I can try to compile without it on MacOSX.

jmborr replied:
> it is because integer has its own `()` operator
What? Integer variables can't be called. Can you give me a code example of an integer using operator `( )`?

wdzhou replied:
What I meant is that std::sort works on integer as if they had operator ().  Even `struct` may not need operator ().

jmborr commented on Framework/MDAlgorithms/src/CompareMDWorkspaces.cpp:
```diff
@@ -31,6 +32,153 @@ class CompareFailsException : public std::runtime_error {
+
+  bool operator()(const SimpleMDEvent &lx, const SimpleMDEvent &rx) const { return lx < rx; }
+
+  bool operator>(const SimpleMDEvent &event2) {
```
Is this ever used?

wdzhou replied:
I don't think so...

jmborr replied:
I would get rid of it. I know it's a pain deleting code after the hard work of writing it but if it's not needed then I think of it as a source of unnecessary complexity.

jmborr commented on Framework/MDAlgorithms/src/CompareMDWorkspaces.cpp:
```diff
@@ -31,6 +32,153 @@ class CompareFailsException : public std::runtime_error {
+    return true;
+  }
+
+  bool operator==(const SimpleMDEvent &event2) {
```
Is this ever used?

wdzhou replied:
I don't think so.

jmborr replied:
same here 🔫

jmborr reviewed (changes requested):
I think a realistic test (system test) with single-crystal diffraction data is needed because I have some misgivings regarding the performance of this multi-dimensional sorting when dealing with millions of events.
This sorting algorithm is exact but I still believe that an approximate sorting (independent sorting of each dimension) will yield correct comparisons for real data in a fraction of the time.

AndreiSavici reviewed (changes requested):
The sorting needs to be parallelized. Needs a test. Too chatty.

AndreiSavici commented:
Try
```python
md = LoadMD(SEQ_11499_md_hwp.nxs', LoadHistory=False)
```
Them compare the workspace with itself. In master it took 2.17 seconds on my computer. Your branch took 28.56 seconds

peterfpeterson committed:
Merge pull request #31823 from mantidproject/SCD364_removeToleranceCheckInSCDCalibratePanels2

Scd364 remove tolerance check in scd calibrate panels2

wdzhou reviewed (commented):
Why 'chatty'?  I don't think that there is anything printed out in sorting method.

`std::sort()` is not parallelized and has O(N log(N)) complexity.   Also building a new vector could be costly.

ckendrick committed:
Fix missing period in doc

rosswhitfield committed:
Merge pull request #31781 from mantidproject/pd273_ikedacarpenter_docs

Update IkedaCarpenterPV doc ornl-next

jmborr commented:
> The sorting needs to be parallelized. Needs a test. Too chatty.

Maybe the parallelization can be carried out for the boxes, [as in here](https://github.com/mantidproject/mantid/blob/SPEC94_imporve_compare_md/Framework/MDAlgorithms/src/ApplyDetailedBalanceMD.cpp#L197)

wdzhou committed:
Refs #105.  Remove 2 unused methods and.

1. Add template for CompareMDWorkspacesTest system test
2. Fix an issue not using abs(diff)

jmborr commented:
> Try to load some systemtest reference data (~800MB)
> 
> ```python
> md = LoadMD('SEQ_11499_md_hwp.nxs', LoadHistory=False)
> ```
> 
> Them compare the workspace with itself. In master it took 2.17 seconds on my computer. Your branch took 28.56 seconds

So comparing events takes about 28.56 - 2.17 = 26.39 s. If the parallelization is successful then comparing events should take just a couple of extra seconds, which sound pretty reasonable 😃

wdzhou committed:
Refs #105. Implementing system test.

wdzhou commented:
There is something wrong with this PR build.  Start a new one!

wdzhou committed:
Refs #105.  Resolved conflict.

wdzhou committed:
Refs #105.  Implemented and passed system test.

wdzhou committed:
Refs #105. Refactored code structure for parallelization-ready.

wdzhou committed:
Refs #105. Enable OpenMP.

wdzhou committed:
Refs #105. Fixed for windows/openmp and add MDTolerance property.

wdzhou committed:
Refs #105.  Started to find how Mac build fails.

wdzhou committed:
Refs #105.  Added back 3 tests.

wdzhou committed:
Refs #105.  Nailed down to 2 tests.

wdzhou committed:
Refs #105.  This shall pass if  CompareMDWorkspaces.

wdzhou committed:
Refs #105.  Disabled CompareMDWorkspaces for test.

wdzhou committed:
Refs #105.  This is supposed to fail.

wdzhou committed:
Refs #105.  Could this be the issue?

wdzhou committed:
Refs #105.  What if I don't check events.

wdzhou committed:
Refs #105.  Turn on event check again.

wdzhou committed:
Refs #105.  Fixed typo.

wdzhou committed:
Refs #105.  Can it be sort?

wdzhou committed:
Refs #105.  Sort is not the reason.

wdzhou commented:
Replaced by https://github.com/mantidproject/mantid/pull/31873 and thus closed without merging.
